src/socket/tcp.rs

```diff
@@ -58,6 +58,9 @@ enum Timer {
     Retransmit {
         expires_at: Instant,
         delay:      Duration
+    }, 
+    FastRetransmit {
+        expires_at: Instant
```

Comment:
I think FastRetransmit doesn't need `expires_at`, since it always fires immediately.

Reply:
Would returning `None` in `poll_at` imply that we want to be polled immidiately?
```rust
fn poll_at(&self) -> Option<Instant> {
    match *self {
        Timer::Idle { keep_alive_at } => keep_alive_at,
        Timer::Retransmit { expires_at, .. } => Some(expires_at),
        Timer::FastRetransmit => None,
        Timer::Close { expires_at } => Some(expires_at),
    }
}
```
https://github.com/barskern/smoltcp/blob/3d4a881b5d4129661a00d32747d6825c80531dc5/src/socket/tcp.rs#L111-L118

Reply:
No, that's an accident. The return value for "polled immediately" is `Some(Instant::from_millis(0))`. You could actually put that in `Retransmit { expires_at }`, but `FastRetransmit` is definitely cleaner.

Reply:
But isn't the return value from `poll_at` supposed to be an absolute value? So we're kinda returning that it should have been polled in the future?

Reply:
I'm not a reviewer, but I think I can answer your question. The return value of `poll_at` lands in `poll_delay` which is used for `wait`. `None` means to block until a packet arrives but a timeout of 0 will return immediately. `poll_delay` will do this for all return values of `poll_at` that are older than the current time given to `poll_delay`. To be sure I think you should just return 0 here as well.

Reply:
See the implementation for UDP socket [here]. `Some(Instant::from_millis(0))` would be some time in the past, which means poll now AFAIK.

[here]: https://github.com/m-labs/smoltcp/blob/master/src/socket/udp.rs#L215-L221

Reply:
(Oh, wrote my answer before all the other comments but did not see them.)

Reply:
Do you think you can make a PR doing that?

Reply:
It's on the way. I'll wait with implementing the rest of this PR until the new `poll_at` is defined.

Reply:
I was thinking it would be appropriate to discuss the signature hence started #215.